File: lib/entity/group_user_profile.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:medicalarm/entity/timestamp.dart';

part 'group_user_profile.g.dart';
part 'group_user_profile.freezed.dart';

/// [Group] 内での [AppUser] のプロファイルを表すエンティティ。
///
/// グループごとに異なる表示名を設定でき、服薬記録の記録者表示などに利用する。
/// Firestore 上では `/groups/{groupID}/userProfiles/` サブコレクションに保存される。
@freezed
abstract class GroupUserProfile with _$GroupUserProfile {
  const GroupUserProfile._();

  @JsonSerializable(explicitToJson: true)
  const factory GroupUserProfile({
    required String id,
    required String groupID,
    required String userID,

    /// グループ内での表示名。
    required String? displayName,
    @ClientCreatedTimestamp() DateTime? createdDateTime,
    @ClientUpdatedTimestamp() DateTime? updatedDateTime,
    @ServerCreatedTimestamp() DateTime? serverCreatedDateTime,
    @ServerUpdatedTimestamp() DateTime? serverUpdatedDateTime,
  }) = _GroupUserProfile;

  factory GroupUserProfile.fromJson(Map<String, dynamic> json) => _$GroupUserProfileFromJson(json);

  /// 複合 ID: `GroupUserProfile_groups_{groupID}_users_{userID}`
  static String documentID({required String groupID, required String userID}) => 'GroupUserProfile_groups_${groupID}_users_$userID';
}

File: firebase/functions/src/entity/group_user_profile.ts
/**
 * GroupUserProfile ドキュメントの複合 ID: GroupUserProfile_groups_{groupID}_users_{userID}
 * クライアント側(lib/entity/group_user_profile.dart の GroupUserProfile.documentID)と同じ規則を維持すること。
 */
export function groupUserProfileDocumentID({
  groupID,
  userID,
}: {
  groupID: string;
  userID: string;
}): string {
  return `GroupUserProfile_groups_${groupID}_users_${userID}`;
}
